### PR TITLE
KAFKA-12432: AdminClient should time out nodes that are never ready

### PR DESCRIPTION
**Pull request by cmccabe**
Previously, if we assigned one or more calls to a remote node, but it
never became available, we would block until the calls hit their the API
timeout.  This was particularly unfortunate in the case where the calls
could have been sent to a different node in the cluster.  This PR fixes
this behavior by timing out pending connections to remote nodes if they
take longer than the request timeout.

There are a few other small cleanups in this PR: it removes the
unecessary Call#aborted, sets Call#curNode to null after the call has
failed to avoid confusion when debugging or logging, and adds a
"closing" boolean rather than setting newCalls to null when the client
is closed.  Also, it increases the log level of the log message that
indicates that we timed out some calls because AdminClient closed.

### CODE REVIEW

**Comment by hachikuji** [comment]
> Previously, we relied solely on the broker to enforce
request.timeout.ms. If the broker failed to enforce this timeout, the
admin client would hang until the call's API timeout hit, which would
return an error to the caller.

I don't think this is accurate. The `NetworkClient` enforces the request timeout, which is passed through the `ClientRequest` object. If the response is not received before the timeout expires, then `NetworkClient` will disconnect.

**Comment by cmccabe** [comment]
> I don't think this is accurate. The NetworkClient enforces the request timeout, which is passed through the ClientRequest object. If the response is not received before the timeout expires, then NetworkClient will disconnect.

Thanks for the correction.  This part of the change description was inaccurate-- I have removed it.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: I wonder if we can raise this to info? This is a really important event to see in the logs, and we are always left guessing about it when the user does not have debug logging enabled. The frequency at the rate of the request timeout should keep it from being too spammy.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:1112]: Maybe I am missing something here. Why do we need to continue? Do we not support multiple in-flight requests? If not, then why does `callsInFlight` map to a list?

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:1129]: It seems like it would be quite intuitive to use `socket.connection.setup.timeout.ms`. Is it because of the complexity of `socket.connection.setup.timeout.max.ms` that we don't do this? Basically we don't know how long the network client itself is planning to wait for the connection to be established.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:1139]: So I guess the intent here is to count the "ready time" toward the request timeout. I think that's reasonable. It means that request timeout represents the maximum time we will commit to trying to get a response from a particular node before retrying. It might be worth a comment to explain this?

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: OK, let's raise it to `INFO`.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:1129]: The complexity of the min / max issue is one thing.  Another thing is that we don't know when the connection has been established, and when it has not.  NetworkClient doesn't expose this information.  `NetworkClient#ready` may return false for a variety of reasons, many of which are not indicative of connection establishment.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:1139]: I added a comment

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:1112]: Notice that we set `maxInFlightRequestsPerConnection` to 1 when constructing the `NetworkClient`.  We don't support sending multiple requests to a single node on a single connection in `AdminClient`.  I think we could add this support, but we'd have to check how the server handled it since we've never done it before.  Maybe there should be a JIRA.

Also, if we do choose to add this support for multiple outstanding requests per node per socket we'd need some way to distinguish between "waiting for a chance to use this connection" from "waiting for this connection to be opened" in NetworkClient.  Currently ready just returns a boolean, which isn't enough information to distinguish these two cases.  We could probably add a new method that returned an enum or something.

Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:1112]: Oh, and on the question of why inflight requests has a list, it was done for future-proofing

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:1112]: Ok. I think we might as well simplify the type since we already have logic which assumes one inflight request. It doesn't feel like we're buying ourselves much in terms of future-proofing.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:1127]: Might be a good idea to use `Math.max(0, deadline - now)` since system time is non-monotonic.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:1153]: I noticed that we lose the throwable that we caught here. Can we add it as a cause?

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java]: While we're here, can we set the exception message to indicate that max retries had been reached?

File: clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java
Comment (hachikuji) [inline clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:5363]: Might be simpler to check `isDone`.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:1112]: ok.  I changed it to just `Map<String, Call>`

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:1127]: Let's just do that in the call to `NetworkClient#poll`.  In general I agree that monotonic time would have been better (maybe this needs a rewrite...)

File: clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java
Comment (cmccabe) [reply clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:5363]: In this case, `isDone` is not supposed to be true until we advance the clock (so that a new node can be tried, after the previous one disconnected)

**Comment by cmccabe** [comment]
For some reason the branch for this PR just stopped changing even when I pushed to it.  Maybe it's a github bug?

I opened a different PR, https://github.com/apache/kafka/pull/10395 , that has the changes we discussed here.

**Review by hachikuji — APPROVED** [review]
LGTM (pending Jenkins approval). Thanks for the patch!

**Pull request merged**
